### PR TITLE
Adding Db to ReName and delete Functions 

### LINKED ISSUE #1: Rename Files Need Db

### PR DESCRIPTION
**Pull request by waqarhussain88**

### CODE REVIEW

File: FileSystemManager/appsettings.json [hunk 1]
```diff
@@ -6,6 +6,6 @@
     }
   },
   "AllowedHosts": "*",
-  "PhotoDirectory": "/Users/mominsalar/Downloads/asda"
+  "PhotoDirectory": "C:\\FileMangar\\Waqar\\Downloads"
```
Comment (FrostyMo) [inline FileSystemManager/appsettings.json]: please dont push this change

File: FileSystemManager/Controllers/FileSystemController.cs [hunk 2]
```diff
@@ -346,98 +357,31 @@ public IActionResult RenameItem(string path, string newName)
...
+                fileMetadata.FileName = newName;
+                fileMetadata.FilePath = newFilePath;
+                fileMetadata.DateModified = DateTime.Now;
+                _context.FileMetadata.Update(fileMetadata);
+                _context.SaveChanges();
 
                 return Ok();
             }
-            catch
+            catch (Exception ex)
```
Comment (FrostyMo) [inline FileSystemManager/Controllers/FileSystemController.cs]: Is there no rollback in case there is an exception?

Comment (FrostyMo) [reply FileSystemManager/Controllers/FileSystemController.cs]: More specifically, code block 

```
if (Directory.Exists(fullPath))
                {
                    Directory.Move(fullPath, newFullPath);
                }
                else if (System.IO.File.Exists(fullPath))
                {
                    System.IO.File.Move(fullPath, newFullPath);
                }
```
Will run and change the filename but db call might fail

**Comment by FrostyMo** [comment]
What about delete function?

**Review by FrostyMo — CHANGES_REQUESTED** [review]

File: FileSystemManager/appsettings.json (hunk 1 above)
Comment (waqarhussain88) [reply FileSystemManager/appsettings.json]: Okay sure

File: FileSystemManager/Controllers/FileSystemController.cs (hunk 2 above)
Comment (waqarhussain88) [reply FileSystemManager/Controllers/FileSystemController.cs]: https://github.com/FrostyMo/FileSystemManager/assets/160615896/fb2e907e-068e-4a1a-8664-86c6e9d59bf8

Please upload the new fine and rename it

File: FileSystemManager/Views/FileSystem/Index.cshtml [hunk 3]
```diff
@@ -353,70 +419,5 @@
             });
         });
     </script>
-    <style>
```
Comment (FrostyMo) [inline FileSystemManager/Views/FileSystem/Index.cshtml]: why is this change shown here? It appears the styles are unchanged?

File: FileSystemManager/Views/Shared/_Layout.cshtml [hunk 4]
```diff
@@ -1,5 +1,6 @@
 ﻿<!DOCTYPE html>
 <html lang="en">
+
```
Comment (FrostyMo) [inline FileSystemManager/Views/Shared/_Layout.cshtml:3]: unnecessary space

File: FileSystemManager/Views/Shared/_Layout.cshtml [hunk 5]
```diff
@@ -8,6 +9,14 @@
     <link rel="stylesheet" href="~/css/site.css" asp-append-version="true" />
     <link rel="stylesheet" href="~/FileSystemManager.styles.css" asp-append-version="true" />
 </head>
+<style>
```
Comment (FrostyMo) [inline FileSystemManager/Views/Shared/_Layout.cshtml]: Why not to add this to site.css in wwwroot?

File: FileSystemManager/appsettings.json (hunk 1 above)
Comment (FrostyMo) [reply FileSystemManager/appsettings.json]: This change is still here

File: FileSystemManager/Controllers/FileSystemController.cs [hunk 6]
```diff
@@ -287,162 +289,249 @@ public IActionResult OpenFile(string path)
...
+                        await _context.SaveChangesAsync();
+                    }
+                    await transaction.CommitAsync();
+                    return Ok();
                 }
-                else
+                catch (Exception ex)
                 {
-                    return NotFound();
+                    await transaction.RollbackAsync();
```
Comment (FrostyMo) [inline FileSystemManager/Controllers/FileSystemController.cs:416]: The transaction fails but directory locally was changed, should it not reverse deletion for local changes?

File: FileSystemManager/Controllers/FileSystemController.cs [hunk 7]
```diff
@@ -287,162 +289,249 @@ public IActionResult OpenFile(string path)
...
                 {
-                    Directory.Move(fullPath, newFullPath);
+                    return NotFound("File not found in recycle bin.");
                 }
-                else if (System.IO.File.Exists(fullPath))
+
+                string deletedPath = fileMetadata.DeletedPath; // Path where the file was originally deleted from
+                string restorePath = DetermineRestorePath(fileMetadata.FilePath); // Determine where to restore the file
+
+                if (System.IO.File.Exists(deletedPath))
```
Comment (FrostyMo) [inline FileSystemManager/Controllers/FileSystemController.cs]: If this is not found here locally, it was still found in DB. What should be done about this inconsistency?

File: FileSystemManager/Controllers/FileSystemController.cs [hunk 8]
```diff
@@ -287,162 +289,249 @@ public IActionResult OpenFile(string path)
...
-        //    //    createdBy = fileMetadata.ModifiedBy,
-        //    //    modifiedDate = fileMetadata.DateModified.ToString("G"),
-        //    //    modifiedBy = fileMetadata.ModifiedBy,
-        //    //    size = FormatSize(fileMetadata.FileSize),
-        //    //    location = relativePath,
-        //    //    expiryDate = fileMetadata.ExpiryDate?.ToString("G"),
-        //    //    issuedBy = fileMetadata.IssuedBy
-        //    //};
-        //    try
+        //    // Decode and normalize the path
```
Comment (FrostyMo) [inline FileSystemManager/Controllers/FileSystemController.cs]: Please remove all commented code

File: FileSystemManager/Controllers/FileSystemController.cs [hunk 9]
```diff
@@ -287,162 +289,249 @@ public IActionResult OpenFile(string path)
...
+                    fileMetadata.FilePath = newFilePath;
+                    fileMetadata.DateModified = DateTime.Now;
+                    _context.FileMetadata.Update(fileMetadata);
+                    _context.SaveChanges();
+                    transaction.Commit();
+                    return Ok();
+                }
+                catch (Exception ex)
+                {
+                    transaction.Rollback();
```
Comment (FrostyMo) [inline FileSystemManager/Controllers/FileSystemController.cs]: Again, it is already renamed locally, what about that rollback?

**Review by FrostyMo — CHANGES_REQUESTED** [review]

File: FileSystemManager/Views/Shared/_Layout.cshtml (hunk 5 above)
Comment (waqarhussain88) [reply FileSystemManager/Views/Shared/_Layout.cshtml]: Done

File: FileSystemManager/Controllers/FileSystemController.cs (hunk 6 above)
Comment (waqarhussain88) [reply FileSystemManager/Controllers/FileSystemController.cs:416]: Added

File: FileSystemManager/Controllers/FileSystemController.cs (hunk 7 above)
Comment (waqarhussain88) [reply FileSystemManager/Controllers/FileSystemController.cs]: Added

File: FileSystemManager/Controllers/FileSystemController.cs (hunk 8 above)
Comment (waqarhussain88) [reply FileSystemManager/Controllers/FileSystemController.cs]: Done

File: FileSystemManager/Controllers/FileSystemController.cs [hunk 10]
```diff
@@ -329,39 +435,81 @@ public IActionResult RenameItem(string path, string newName)
...
+            }
+
+            var newFullPath = Path.Combine(Path.GetDirectoryName(fullPath), newName).Replace('\\', '/');
+
+            using (var transaction = _context.Database.BeginTransaction())
+            {
+                try
                 {
-                    System.IO.File.Move(fullPath, newFullPath);
+                    bool itemExistsLocally = false;
```
Comment (FrostyMo) [inline FileSystemManager/Controllers/FileSystemController.cs]: why use this if we can return NotFound() in else condition if not found?

File: FileSystemManager/Controllers/FileSystemController.cs [hunk 11]
```diff
@@ -329,39 +435,81 @@ public IActionResult RenameItem(string path, string newName)
...
+                        return NotFound("Item not found on the local file system. Metadata marked as deleted.");
+                    }
+
+                    return Ok();
                 }
-                else
+                catch (Exception ex)
                 {
-                    return NotFound();
+                    transaction.Rollback();
```
Comment (FrostyMo) [inline FileSystemManager/Controllers/FileSystemController.cs]: This is still left

File: FileSystemManager/Controllers/FileSystemController.cs [hunk 12]
```diff
@@ -286,41 +275,158 @@ public IActionResult OpenFile(string path)
...
+        public async Task<IActionResult> PermanentlyDeleteItem(int id)
+        {
+            try
+            {
+                var fileMetadata = await _context.FileMetadata.FindAsync(id);
+                if (fileMetadata == null || !fileMetadata.IsDeleted)
+                {
+                    return NotFound();
+                }
+                if (System.IO.File.Exists(fileMetadata.DeletedPath))
```
Comment (FrostyMo) [inline FileSystemManager/Controllers/FileSystemController.cs]: No directory check?

File: FileSystemManager/Controllers/FileSystemController.cs [hunk 13]
```diff
@@ -286,41 +275,158 @@ public IActionResult OpenFile(string path)
...
+                    if (Directory.Exists(fullPath))
+                    {
+                        Directory.Move(fullPath, recycleBinPath);
+                        isDirectory = true;
+                    }
+                    else if (System.IO.File.Exists(fullPath))
+                    {
+                        if (System.IO.File.Exists(recycleBinPath))
+                        {
+                            System.IO.File.Delete(recycleBinPath);
```
Comment (FrostyMo) [inline FileSystemManager/Controllers/FileSystemController.cs:387]: why is recyclebinpath in normal delete function?

File: FileSystemManager/Controllers/FileSystemController.cs (hunk 7 above)
Comment (FrostyMo) [reply FileSystemManager/Controllers/FileSystemController.cs]: What about directory?

**Review by FrostyMo — CHANGES_REQUESTED** [review]

File: FileSystemManager/Controllers/FileSystemController.cs (hunk 10 above)
Comment (waqarhussain88) [reply FileSystemManager/Controllers/FileSystemController.cs]: why use this if we can return NotFound() in else condition if not found? Discuss about this condition is true 
now we need to same condition on the db level if any person delete manullay what should we do In that case

File: FileSystemManager/Controllers/FileSystemController.cs [hunk 14]
```diff
@@ -480,101 +869,12 @@ private List<Breadcrumb> GenerateBreadcrumbs(string path)
...
-
         public class Breadcrumb
         {
             public string Name { get; set; }
             public string Path { get; set; }
         }
-
-
-        [HttpPost]
-        public IActionResult DownloadSelected([FromBody] List<string> selectedFiles)
```
Comment (FrostyMo) [inline FileSystemManager/Controllers/FileSystemController.cs:496]: Why is this removed?

File: FileSystemManager/Controllers/FileSystemController.cs [hunk 15]
```diff
@@ -480,101 +869,12 @@ private List<Breadcrumb> GenerateBreadcrumbs(string path)
...
-                    }
-                }
-
-                memoryStream.Seek(0, SeekOrigin.Begin);
-                return File(memoryStream.ToArray(), "application/zip", "selected_files.zip");
-            }
-        }
-
-        // Helper method to add a directory and its contents to a zip file
-        private void AddDirectoryToZip(ZipArchive archive, string sourceDir, string entryName)
```
Comment (FrostyMo) [inline FileSystemManager/Controllers/FileSystemController.cs:530]: And this?

File: FileSystemManager/Controllers/FileSystemController.cs [hunk 16]
```diff
@@ -480,101 +869,12 @@ private List<Breadcrumb> GenerateBreadcrumbs(string path)
...
-
-            // Handle the case for empty directories
-            if (files.Length == 0 && directories.Length == 0)
-            {
-                archive.CreateEntry(Path.Combine(entryName, "empty_directory.txt").Replace("\\", "/"));
-            }
-        }
-
-        [HttpPost]
-        public IActionResult DeleteSelected([FromBody] List<string> paths)
```
Comment (FrostyMo) [inline FileSystemManager/Controllers/FileSystemController.cs:554]: And this?

**Review by FrostyMo — CHANGES_REQUESTED** [review]

File: FileSystemManager/Controllers/FileSystemController.cs (hunk 15 above)
Comment (waqarhussain88) [reply FileSystemManager/Controllers/FileSystemController.cs:530]: YES

File: FileSystemManager/Views/Shared/_Layout.cshtml (hunk 4 above)
Comment (waqarhussain88) [reply FileSystemManager/Views/Shared/_Layout.cshtml:3]: Removed

**Review by FrostyMo — APPROVED** [review]
